cilium: backport v1.11 loopback fix

jrfastab opened the pull request:
Backport fix

Commits:
 2c9c8c17aeeb bpf: Fix LB loopback path with ingress policy (#22972)
 dbf4351c4eff1 bpf: lxc: defer CT_INGRESS entry creation for loopback connections (#27602)
 e2829a061a53 bpf: lxc: support Pod->Service->Pod hairpinning with endpoint routes (#27798)

---

Once this PR is merged, you can update the PR labels via:
```upstream-prs
for pr in 22972 27602 27798; do contrib/backporting/set-labels.py $pr done 1.11; done
```
or with
```
make add-labels BRANCH=v1.11 ISSUES=22972,27602,27798
```

aditighag committed:
bpf: Fix LB loopback path with ingress policy

[ upstream commit 2c9c8c17aeebcd2e295ecd29fd99305000bc7790 ]

When a service backend endpoint connects to itself via its service cluster IP
address, the traffic is hairpin'd using a loopback IP address. We skip policy
enforcement on egress, as a pod is allowed to connect to itself, and users don't
have to specify additional policies to allow the hairpin'd traffic. We have a
similar expectation on the ingress side.
The ingress side was broken because of a regression due to which replies were
dropped on the ingress policy enforcement path. The patch that introduced the
regression split per packet load-balancing logic into a separate tail
call, where (limited) LB state is stored in packet context, and restored later
while handling rest of the packet processing including conntrack, and policy
enforcement.
When a service pod connects to itself via its clusterIP, post the forward
service translation, conntrack state update is done on the conntrack entry
with loopback address (the restored state). As a result, when a reply is
reverse SNAT'd and DNAT'd, the policy verdict for the `CT_NEW` entry is
denied, and the reply packet is dropped.  Prior to the regression, the
original conntrack entry would have the updated state, and loopback flag set.

Fix: When hairpin'd traffic is reverse translated, and sent for a local delivery
in case of bpf_lxc, we should directly redirect it to the endpoint, thereby
skipping the tail call with ingress policy enforcement altogether.

Here is the packet flow -

Request path: podA -> clusterIP --> svc xlate (dAddr=podA) --> SNAT using
loopback address (saddr=loopback IP) --> conntrack entry created (loopback IP,
backend IP) with loopback flag state

Reply path: cluster IP -> podA --> svc rev-xlate (saddr=podA address) --> SNAT
using loopback (daddr=loopback IP) --> CT_REPLY match --> local_delivery

[ Backport note: Addressed some minor conflicts. ]

Fixes: 7575ba03cc ("datapath: Reduce from LXC complexity")
Signed-off-by: Aditi Ghag <aditi@cilium.io>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

julianwiedmann committed:
bpf: lxc: defer CT_INGRESS entry creation for loopback connections

[ upstream commit dbf4351c4eff1857cf374886804eb876e970b992 ]

Currently the CT_INGRESS entry for a loopback connection is already created
when the first request leaves the client, as from-container creates the
CT_EGRESS entry (see the loopback handling in ct_create4() for details).

This is unusual - we would normally just create the CT_INGRESS entry as
the first packet passes through to-container into the backend. But for
loopback connections it is needed, so that
1.) to-container finds a CT entry with .loopback set, and thus skips
    network policy enforcement even for the first packet, and
2.) the CT entry has its rev_nat_index field populated, and thus can
    RevNAT replies in from-container.

This approach conflicts with the fact that loopback replies skip the
client's to-container path (to avoid network policy enforcement).

Once the loopback connection is closed, the backend's from-container path
observes the FIN / RST, and __ct_lookup4() updates the CT_INGRESS entry's
lifetime to CT_CLOSE_TIMEOUT. But the client's to-container path will not
observe the FIN / RST, and thus the CT_EGRESS entry's lifetime remains as
CT_CONNECTION_LIFETIME_TCP. Therefore the CT_INGRESS entry will expire
earlier, and potentially gets garbage-collected while the CT_EGRESS entry
stays in place.

If the same loopback connection is subsequently re-opened, the client's
from-container path finds the CT_EGRESS entry and thus will *not* call
ct_create4(). Consequently the CT_INGRESS entry is not created, and the
backend will not apply the loopback-specific handling described above.
Inbound requests are potentially dropped (due to network policy), and/or
replies are not RevNATed.

Fix this up by letting the backend path create its CT_INGRESS entry as
usual. It just needs a bit of detection code in its CT_NEW handling to
understand that the first packet belongs to a .loopback connection, and
populate its own CT_INGRESS entry accordingly.

[ Backport note: Addressed some ~minor~ conflicts, skipped tests. ]

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

ti-mo committed:
bpf: lxc: support Pod->Service->Pod hairpinning with endpoint routes

[ upstream commit e2829a061a53e5667d22bc9159ddc17038b45dfb ]

This patch adds a check to determine whether a Pod is sending traffic to
itself through a Service. With endpoint routes enabled, this traffic is
subject to policy, and we want to avoid the user to have to explicitly
allow traffic to+from itself.

[ Backport note:
    Fix conflicts, in particular add lb4_extract_tuple(),
    missing on v1.12 and v1.11].

Signed-off-by: Timo Beckers <timo@isovalent.com>
Co-authored-by: Quentin Monnet <quentin@isovalent.com>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

Commit:
Revert "endpoint: Do not override deny entries with proxy redirects"

This reverts commit fefa16b03ab2e1ce8b208be3252a4b7bcfb3d9c3.

Signed-off-by: Marga Manterola <marga@isovalent.com>

Commit:
Revert "policy: Export DenyPreferredInsertWithChanges, make revertible"

This reverts commit 6dc3fb3d137af367a5a75245257f848e5a0c437e.

Signed-off-by: Marga Manterola <marga@isovalent.com>

Commit:
Revert "bug: Fix Potential Nil Reference in GetLables Implementation"

This reverts commit 7dc319e4e8e5c7e1d42d4e2f8c0fbc364bcd6b60.

Signed-off-by: Marga Manterola <marga@isovalent.com>

Commit:
Revert "policy: Fix Deny Precedence Bug"

This reverts commit b6970da61cf40b6dccc92e6ce60369c195c9a21e.

Except for the logrus update.

Signed-off-by: Marga Manterola <marga@isovalent.com>

qmonnet commented on bpf/bpf_lxc.c:
```diff
@@ -778,7 +789,8 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
 			return ipv4_local_delivery(ctx, l3_off, SECLABEL, ip4,
-						   ep, METRIC_EGRESS, false);
+						   ep, METRIC_EGRESS,
+						   bypass_ingress_policy);
```
This does not seem correct, we're changing the `from_host` argument instead of `hairpin_flow`, which was added in fcbd5d780bc5085f4cc24f197abaab0be2989143.

I'm unsure whether we should also backport this other commit or not. My understanding is that it should address a similar issue to what we're trying to fix with the current backport, but when per-endpoint routes are _not_ enabled. I suppose that without a compelling reason, we should avoid pulling more commits for this branch, so I'd just drop this hunk (and the previous one where we initialise `bypass_ingress_policy`).

Cc @aditighag FYI.

jrfastab replied:
removed that hunk and this change. @qmonnet look good to you now?

qmonnet reviewed (approved):
Haven't been able to test this afternoon because I couldn't get my 1.11 dev setup to work anymore :facepalm: but the backport does look correct now, thank you!

qmonnet committed:
Remove no more available dockershim flags in kubelet wrapper

Since k8s v1.24 dockershim related flags like `--network-plugin` and
`--cni-bin-dir` have been removed. Keeping them leads to the kubelet
exiting with an error.

When Cilium is deployed on GKE it needs to substitute the real kubelet
executable with a wrapper script, in order to stop containerd and remove
any existing CNI configuration files before starting the kubelet itself.

With this commit, the startup script in cilium-node-init first checks
the kubelet version and, before starting it, adds the now unavailable
options only if k8s version is below v1.24.

Besides, since GKE nodes with v1.24 use containerd v1.6, a missing CNI
config is no more allowed, leading to containerd not responding to the
kubelet in such case.  Thus, the wrapper script also creates a dummy CNI
config before restarting containerd.

Co-authored-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
Co-authored-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>
